### PR TITLE
[CM-402]Add the support text validation in form template

### PR DESCRIPTION
**Pull request by Sunilkumarr**
## Summary
Added the text form validation this on top of CM-402 https://github.com/AppLozic/ApplozicSwift/pull/361 

The form validation will work based on the validation in json for text type and it will have regex and error message to show in the label. 

Once the user clicks on the submit button we will validate and show the error message in case if it's not valid below the text fields.

Form Error label when the email is not valid :

Simulator Screen Shot - iPhone 11 - 2020-10-06 at 18 29 17

Form when the email is valid :

Simulator Screen Shot - iPhone 11 - 2020-10-06 at 18 29 38

## Testing
Tested by sending json from postman for checking form data validation for text type.

JSON : 

`{
 "groupId" : "46383008",
 "message": "Hi How are you",
"metadata": { "contentType": "300", "templateId": "12", "payload": "[ { \"type\": \"text\", \"data\": { \"label\": \"Name\", \"placeholder\": \"Enter your name\" } }, { \"data\": { \"placeholder\": \"Enter Password\", \"label\": \"Password\" }, \"type\": \"password\" }, { \"data\": { \"placeholder\": \"Enter your Email ID\", \"label\": \"emailId\", \"validation\": { \"regex\": \"[a-z0-9._%+-]+@[a-z0-9.-]+\\\\.[a-z]{2,3}\", \"errorText\": \"Invalid Email\" } }, \"type\": \"text\" }, { \"data\": { \"label\": \"employee Type\", \"placeholder\": \"Enter your Employee Type\" }, \"type\": \"text\" }, { \"data\": { \"value\": \"some_user_secrets\", \"name\": \"userDetail\" }, \"type\": \"hidden\" }, { \"data\": { \"options\": [ { \"value\": \"male\", \"label\": \"Male\" }, { \"label\": \"Female\", \"value\": \"female\" }, { \"label\": \"Other\", \"value\": \"other\" } ], \"title\": \"What do you identify as:\", \"name\": \"Gender\" }, \"type\": \"radio\" }, { \"type\": \"checkbox\", \"data\": { \"title\": \"Choose your Music Type:\", \"name\": \"music_type_b\", \"options\": [ { \"value\": \"metal\", \"label\": \"Metal\" }, { \"value\": \"pop\", \"label\": \"Pop\" }, { \"label\": \"Blues\", \"value\": \"blues\" }, { \"value\": \"folk\", \"label\": \"Folk\" } ] } }, { \"type\": \"radio\", \"data\": { \"title\": \"What do you fav sports:\", \"options\": [ { \"value\": \"cricket\", \"label\": \"Cricket\" }, { \"label\": \"FootBall\", \"value\": \"football\" }, { \"label\": \"Cards\", \"value\": \"cards\" }, { \"label\": \"Chess\", \"value\": \"chess\" } ], \"name\": \"sports\" } }, { \"data\": { \"title\": \"Choose your dance Type:\", \"options\": [ { \"label\": \"Jazz\", \"value\": \"jazz\" }, { \"label\": \"Classical\", \"value\": \"clasical\" }, { \"label\": \"Lyrical\", \"value\": \"lyrical\" }, { \"label\": \"Zumba\", \"value\": \"zumba\" } ], \"name\": \"music_type_a\" }, \"type\": \"checkbox\" }, { \"data\": { \"name\": \"Submit\", \"action\": { \"formAction\": \"https:\/\/reqres.in\/api\/users\", \"requestType\": \"json\", \"message\": \"optional- this message will be used as acknowledgement text when user clicks the button\" } }, \"type\": \"submit\" } ]"
  }
}`

### CODE REVIEW

File: Sources/Views/ALKFormCell.swift [hunk 1]
```diff
@@ -331,6 +352,60 @@ extension ALKFormCell: ALKDatePickerButtonClickProtocol {
...
+        else {
+            return false
+        }
+
+        for index in 0 ..< viewModelItems.count {
+            let element = viewModelItems[index]
+
+            switch element.type {
+            case .text:
+                var enterdText = ""
```
Comment (mukeshthawani) [inline Sources/Views/ALKFormCell.swift:370]: Typo: "entered"

File: Sources/Views/ALKFormCell.swift [hunk 2]
```diff
@@ -331,6 +352,60 @@ extension ALKFormCell: ALKDatePickerButtonClickProtocol {
...
+                let textFieldModel = element as? FormViewModelTextItem
+
+                if formDataSubmit.textFields.isEmpty {
+                    enterdText = ""
+                } else {
+                    enterdText = formDataSubmit.textFields[index] ?? ""
```
Comment (mukeshthawani) [inline Sources/Views/ALKFormCell.swift:376]: `isEmpty` check is not required, this is enough: `enteredText = formDataSubmit.textFields[index] ?? ""`

File: Sources/Views/ALKFormCell.swift [hunk 3]
```diff
@@ -331,6 +352,60 @@ extension ALKFormCell: ALKDatePickerButtonClickProtocol {
...
+                    let regxPattern = validation.regex
+                {
+                    let range = NSRange(enterdText.startIndex ..< enterdText.endIndex, in: enterdText)
+
+                    do {
+                        let regex = try NSRegularExpression(pattern: regxPattern, options: [])
+                        let matches = regex.numberOfMatches(in: enterdText, options: [], range: range)
+                        print("Form data text matches: \(matches)")
+                        if matches > 0 {
+                            formDataSubmit.validationFields[index] = 1
```
Comment (mukeshthawani) [inline Sources/Views/ALKFormCell.swift:389]: We can move the validation part to a separate validator class.

File: Sources/Views/ALKFormTextItemCell.swift [hunk 4]
```diff
@@ -34,6 +34,24 @@ class ALKFormTextItemCell: UITableViewCell {
...
+        labelStackView.distribution = .fillEqually
+        labelStackView.backgroundColor = UIColor.white
+        return labelStackView
+    }()
+
+    let errorLabel: UILabel = {
+        let label = UILabel(frame: .zero)
+        label.textColor = .red
+        label.font = Font.normal(size: 15).font()
+        label.text = "This is error lol"
```
Comment (mukeshthawani) [inline Sources/Views/ALKFormTextItemCell.swift:50]: Update this text and move it to localization if possible.

File: Sources/Controllers/ALKConversationViewController+TableView.swift [hunk 5]
```diff
@@ -516,10 +516,34 @@ extension ALKConversationViewController: UITableViewDelegate, UITableViewDataSou
...
                     guard let weakSelf = self else { return }
-                    weakSelf.formSubmitButtonSelected(formSubmitData: submitData,
-                                                      messageModel: message,
-                                                      isButtonClickDisabled:
-                                                      weakSelf.configuration.disableRichMessageButtonAction)
+
+                    // If not valid reload the table view section for a form to show the error message below the text fields.
+                    if !cell.isFormDataValid() {
+                        guard let index = weakSelf.viewModel.sectionFor(identifier: message.identifier),
+                            index < weakSelf.tableView.numberOfSections
```
Comment (mukeshthawani) [inline Sources/Controllers/ALKConversationViewController+TableView.swift:523]: Move the reloading and section validation to a separate function as we're doing this at multiple places.

File: Sources/Views/ALKFormCell.swift [hunk 6]
```diff
@@ -331,6 +352,60 @@ extension ALKFormCell: ALKDatePickerButtonClickProtocol {
...
+            let viewModelItems = template?.viewModeItems
+        else {
+            return false
+        }
+
+        for index in 0 ..< viewModelItems.count {
+            let element = viewModelItems[index]
+
+            switch element.type {
+            case .text:
```
[the remark this answers is not in the document]
Comment (Sunilkumarr) [reply Sources/Views/ALKFormCell.swift:374]: Checking all the text types in the form.

File: Sources/Views/ALKFormCell.swift (hunk 1 above)
Comment (Sunilkumarr) [reply Sources/Views/ALKFormCell.swift]: corrected

File: Sources/Views/ALKFormCell.swift (hunk 2 above)
Comment (Sunilkumarr) [reply Sources/Views/ALKFormCell.swift]: changed

File: Sources/Views/ALKFormCell.swift (hunk 3 above)
Comment (Sunilkumarr) [reply Sources/Views/ALKFormCell.swift]: moved

File: Sources/Views/ALKFormTextItemCell.swift (hunk 4 above)
Comment (Sunilkumarr) [reply Sources/Views/ALKFormTextItemCell.swift]: moved to localization

File: Sources/Controllers/ALKConversationViewController+TableView.swift (hunk 5 above)
Comment (Sunilkumarr) [reply Sources/Controllers/ALKConversationViewController+TableView.swift]: Moved

**Review by mukeshthawani — APPROVED** [review]

**Pull request merged**